MobiusCore/Source/EffectHandlers/EffectExecutor.swift

```diff
@@ -0,0 +1,100 @@
+// Copyright (c) 2019 Spotify AB.
+//
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+import Foundation
+
+class EffectExecutor<Effect, Event>: Connectable {
+    private let handleEffect: (Effect, Response<Event>) -> Disposable
+    private var output: Consumer<Event>?
+
+    private let lock = Lock()
+
+    // Keep track of each received effect's state.
+    // When an effect has completed, it should be removed from this dictionary.
+    // When disposing this effect handler, all entries must be removed.
+    private var handlingEffects: [Int64: EffectHandlingState<Event>] = [:]
+    private var nextID = Int64(0)
+
+    init(handleInput: @escaping (Effect, Response<Event>) -> Disposable) {
+        self.handleEffect = handleInput
+    }
+
+    func connect(_ consumer: @escaping Consumer<Event>) -> Connection<Effect> {
+        output = consumer
```

Comment:
This should check if we're already connected to something - otherwise, there's a risk that the new consumer gets events intended for the previous one via the `output`.

Reply:
(Should be easy to guard with a test, too)